### PR TITLE
Issue #126 Copy-GcsObject

### PR DESCRIPTION
**Pull request by ILMTitan**
Added Copy-GcsObject.
#126

### CODE REVIEW

File: Google.PowerShell/Common/GcloudCmdlet.cs [hunk 1]
```diff
@@ -43,7 +43,7 @@ public GCloudCmdlet()
         /// <summary>
         /// Returns an instance of the Google Client API initializer, using the machine's default credentials.
         /// </summary>
-        protected BaseClientService.Initializer GetBaseClientServiceInitializer()
+        protected static BaseClientService.Initializer GetBaseClientServiceInitializer()
```
Comment (chrsmith) [inline Google.PowerShell/Common/GcloudCmdlet.cs]: Can you back this change out? It seems unrelated to creating a new `Copy-GcsObject` cmdlet, and we should talk about the design implications here.

File: Google.PowerShell/Storage/GcsCmdlet.cs [hunk 2]
```diff
@@ -24,12 +24,10 @@ public abstract class GcsCmdlet : GCloudCmdlet
...
         protected const string UTF8TextMimeType = "text/plain; charset=utf-8";
 
-        // TODO(chrsmith): Cache the storage service? Create it in OnProcessRecord every time? (So it does so once?)
-
-        protected StorageService GetStorageService()
-        {
-            return new StorageService(GetBaseClientServiceInitializer());
-        }
+        /// <summary>
+        /// The storage service.
```
Comment (chrsmith) [inline Google.PowerShell/Storage/GcsCmdlet.cs]: I'm not a huge fan of this, but don't have a good reason to oppose the change.

The initial design was so that we would explicitly control the lifetime of the service object. That would make it easier to mock, test, instrument, etc.

However, we could still probably do all of those same things by going through a property accessor. And this approach also cleans up a lot of the code.

File: Google.PowerShell/Storage/GcsObject.cs [hunk 3]
```diff
@@ -922,4 +917,75 @@ protected override void ProcessRecord()
             }
         }
     }
+
+    [Cmdlet(VerbsCommon.Copy, "GcsObject", DefaultParameterSetName = ParameterSetNames.ByObject)]
+    [OutputType(typeof(Object))]
+    public class CopyGcsObject : GcsCmdlet
```
Comment (chrsmith) [inline Google.PowerShell/Storage/GcsObject.cs]: Know what I could go for? Some XML doc comments. A delicious "summary" and "description" tag would be great right about now.

File: Google.PowerShell/Storage/GcsObject.cs [hunk 4]
```diff
@@ -922,4 +917,75 @@ protected override void ProcessRecord()
...
+    {
+
+        private class ParameterSetNames
+        {
+            public const string ByName = "ByName";
+            public const string ByObject = "ByObject";
+        }
+
+        [Parameter(ParameterSetName = ParameterSetNames.ByObject, Mandatory = true, ValueFromPipeline = true)]
+        public Object InputObject { get; set; }
```
Comment (chrsmith) [inline Google.PowerShell/Storage/GcsObject.cs]: What does this parameter do?

File: Google.PowerShell/Storage/GcsObject.cs [hunk 5]
```diff
@@ -922,4 +917,75 @@ protected override void ProcessRecord()
...
+        /// Name of the bucket containing the object. Will also accept a Bucket object.
+        /// </para>
+        /// </summary>
+        [Parameter(ParameterSetName = ParameterSetNames.ByName, Mandatory = true)]
+        [PropertyByTypeTransformation(Property = nameof(Bucket.Name), TypeToTransform = typeof(Bucket))]
+        public string SourceBucket { get; set; }
+
+        /// <summary>
+        /// <para type="description">
+        /// Name of the object to write to.
```
Comment (chrsmith) [inline Google.PowerShell/Storage/GcsObject.cs]: Name of the object to write to? On `SourceObjectName`? Do you mean to write "Name of the object to read from?

File: Google.PowerShell/Storage/GcsObject.cs [hunk 6]
```diff
@@ -922,4 +917,75 @@ protected override void ProcessRecord()
...
+        /// Name of the bucket in which the copy will reside.
+        /// </para>
+        /// </summary>
+        [Parameter(Mandatory = false, Position = 0)]
+        [PropertyByTypeTransformation(Property = nameof(Bucket.Name), TypeToTransform = typeof(Bucket))]
+        public string DestinationBucket { get; set; }
+
+        /// <summary>
+        /// <para type="description">
+        /// The name of the copy.
```
Comment (chrsmith) [inline Google.PowerShell/Storage/GcsObject.cs]: Perhaps "Name of the newly created Object" would be clearer?

File: Google.PowerShell.IntegrationTests/Storage/Storage.GcsObject.Tests.ps1 [hunk 7]
```diff
@@ -600,4 +600,41 @@ Describe "Test-GcsObject" {
...
+            Copy-GcsObject -SourceBucket $bucket -SourceObject "test-source" $bucket "test-dest"
+            Read-GcsObject $bucket "test-dest" | Should Be "test1-$r"
+        }
+
+        It "Should work by object" {
+            $sourceObj = New-GcsObject $bucket "test-source2" -Contents "test2-$r"
+            $sourceObj | Copy-GcsObject $bucket "test-dest2"
+            Read-GcsObject $bucket "test-dest2" | Should Be "test2-$r"
+        }
+
```
Comment (chrsmith) [inline Google.PowerShell.IntegrationTests/Storage/Storage.GcsObject.Tests.ps1]: What if you don't have permissions to read from the source bucket? Or write to the destination bucket? I imagine there is a test we need here.

File: Google.PowerShell.IntegrationTests/Storage/Storage.GcsObject.Tests.ps1 [hunk 8]
```diff
@@ -600,4 +600,41 @@ Describe "Test-GcsObject" {
...
+        AfterAll {
```
Comment (chrsmith) [inline Google.PowerShell.IntegrationTests/Storage/Storage.GcsObject.Tests.ps1]: What if a GCS object already exists in the destination? I assume we need an `-Overwrite` or `-Force` flag?

File: Google.PowerShell/Storage/GcsObject.cs [hunk 9]
```diff
@@ -922,4 +917,75 @@ protected override void ProcessRecord()
...
+                    break;
+                case ParameterSetNames.ByObject:
+                    gcsObject = InputObject;
+                    break;
+                default:
+                    throw UnknownParameterSetException;
+            }
+            ObjectsResource.CopyRequest request = Service.Objects.Copy(gcsObject,
+                gcsObject.Bucket, gcsObject.Name,
+                DestinationBucket ?? gcsObject.Bucket, DestinationObjectName ?? gcsObject.Name);
```
Comment (chrsmith) [inline Google.PowerShell/Storage/GcsObject.cs]: Having defaults for these values seems reasonable. But you should mention that in the doc comments above.

**Comment by chrsmith** [comment]
The code generally looks good. A few minor issues.

**Comment by ILMTitan** [comment]
PTAL

File: Google.PowerShell/Storage/GcsObject.cs [hunk 10]
```diff
@@ -939,14 +939,15 @@ private class ParameterSetNames
         }
 
         /// <summary>
-        /// A Google Cloud Storage object description 
+        /// A Google Cloud Storage object description of the object to read from. They can be retrieved using
```
Comment (chrsmith) [inline Google.PowerShell/Storage/GcsObject.cs]: Needs `<para type="description">`.

Comment (chrsmith) [inline Google.PowerShell/Storage/GcsObject.cs]: "object description of the object to read from" sounds confusing. How about just

"Cloud Storage object to read from. Can be obtained using Get-GcsObject or Find-GcsObject."

File: Google.PowerShell/Storage/GcsObject.cs [hunk 11]
```diff
@@ -993,9 +1000,30 @@ protected override void ProcessRecord()
...
+
+            if (!Force)
+            {
+                try
+                {
+                    ObjectsResource.GetRequest objGetReq =
+                        Service.Objects.Get(destinationBucket, destinationObject);
+                    objGetReq.Execute();
+                    // If destination does not exist, jump to catch statment.
+                    if (!ShouldContinue("Object exists. Overwrite?",
```
Comment (chrsmith) [inline Google.PowerShell/Storage/GcsObject.cs]: If this is too long for a single line, line break at the first parameter of `ShouldContinue`, otherwise this looks kinda odd IMHO.

File: Google.PowerShell/Storage/GcsObject.cs [hunk 12]
```diff
@@ -993,9 +1000,30 @@ protected override void ProcessRecord()
...
+            string destinationObject = DestinationObjectName ?? gcsObject.Name;
+
+            if (!Force)
+            {
+                try
+                {
+                    ObjectsResource.GetRequest objGetReq =
+                        Service.Objects.Get(destinationBucket, destinationObject);
+                    objGetReq.Execute();
+                    // If destination does not exist, jump to catch statment.
```
Comment (chrsmith) [inline Google.PowerShell/Storage/GcsObject.cs]: "jump to catch statement" sounds like you are using a goto. Perhaps:

"If destination does not exist, will throw a 404 an be ignored"

File: Google.PowerShell.IntegrationTests/Storage/Storage.GcsObject.Tests.ps1 [hunk 13]
```diff
@@ -618,6 +618,17 @@ Describe "Copy-GcsObject" {
...
+        It "Should fail to read from unaccessable source bucket" {
+            { Copy-GcsObject -SourceBucket "asdf" -SourceObject "test-source" $bucket "test-dest" } |
+                Should Throw 403
+        }
+        
+        It "Should fail to write to unaccessable source bucket" {
+            New-GcsObject $bucket "test-source0" -Contents "test0-$r"
+            { Copy-GcsObject -SourceBucket $bucket -SourceObject "test-source0" "asdf" "test-dest" } |
+                Should Throw 403
+        }
```
Comment (chrsmith) [inline Google.PowerShell.IntegrationTests/Storage/Storage.GcsObject.Tests.ps1]: Need a test to validate `-Force works as expected`. just copy your "should work by name" test, and add another scenario where it fails (already exists) and then works via `-Force`.

**Comment by chrsmith** [comment]
LGTM, but please add the missing test cases and XML doc comment. Feel free to pass on suggested wordings for doc strings.

File: Google.PowerShell/Storage/GcsObject.cs (hunk 12 above)
Comment (ILMTitan) [inline Google.PowerShell/Storage/GcsObject.cs]: It sounds like I am using a goto because I basically am using a goto.

**Comment by ILMTitan** [comment]
Close #91

File: Google.PowerShell.IntegrationTests/Storage/Storage.GcsObject.Tests.ps1 [hunk 14]
```diff
@@ -642,6 +642,12 @@ Describe "Copy-GcsObject" {
             Read-GcsObject $bucket "test-dest2" | Should Be "test2-$r"
         }
 
+        It "Should overwrite existing object" {
+            $sourceObj = Get-GcsObject $bucket "test-source"
+            $sourceObj | Copy-GcsObject $bucket "test-dest2" -Force
```
Comment (chrsmith) [inline Google.PowerShell.IntegrationTests/Storage/Storage.GcsObject.Tests.ps1]: Sorry to be a pain here, but could you confirm an error if you don't specify `-Force`? Otherwise it could be that we always overwrite, regardless of if you set the flag.

````
{ $sourceObj | Copy-GcsObject $bucket "should-fail" } |
    Should Throw 409  # Conflit
````

**Comment by chrsmith** [comment]
LGTM

File: Google.PowerShell.IntegrationTests/Storage/Storage.GcsObject.Tests.ps1 (hunk 14 above)
Comment (ILMTitan) [inline Google.PowerShell.IntegrationTests/Storage/Storage.GcsObject.Tests.ps1]: I can not. I don't throw an error. Instead I request confirmation from the user. However, the pester story for testing "ShouldConfirm" does not yet exist.

Comment (chrsmith) [inline Google.PowerShell.IntegrationTests/Storage/Storage.GcsObject.Tests.ps1]: Gotcha. All is well then. :+1:
